File: src/main/java/com/kshitij/pocs/scloud/product/service/ProductService.java
package com.kshitij.pocs.scloud.product.service;

import com.kshitij.pocs.scloud.product.entity.Product;
import com.kshitij.pocs.scloud.product.exceptions.ProductNotFoundException;
import com.kshitij.pocs.scloud.product.repository.ProductRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ProductService {


    private ProductRepository productRepository;
    private MessageProducerService messageProducerService;


    public ProductService(ProductRepository productRepository,MessageProducerService messageProducerService){
        this.productRepository=productRepository;
        this.messageProducerService=messageProducerService;
    }

    public Product getProductById(String id) {
        return this.productRepository.findById(id).orElseThrow(ProductNotFoundException::new);
    }
    public Iterable<Product> getAllProducts(){
        return this.productRepository.findAll();
    }
    public Product createProduct(Product product){
        Product createdProduct = this.productRepository.save(product);
        this.messageProducerService.sendMessage(product,Boolean.FALSE,"product.update");
        return createdProduct;
    }
    public List<Product> getProductByCategory(Long categoryId){
        return this.productRepository.getProductByCategory(categoryId);
    }

    public Product updateProduct(String id, Product product) {
        //Product existingProd=this.productRepository.findById(id);
        Product updatedProduct =this.productRepository.save(product);
        this.messageProducerService.sendMessage(updatedProduct,Boolean.FALSE,"product.update");
        return updatedProduct;
    }
}

File: src/main/kotlin/kshitij/pocs/scloud/productservice/ProductService.kt
package kshitij.pocs.scloud.productservice

import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

/*
TODO
1. Serialize message as JSON and then send to rabbit MQ
2. Deserialise message and then Recieve in Product Search Service
 */

@Service
class ProductService(val productRepository: ProductRepository, val messageProducerService: MessageProducerService){
    fun getProductById(id:String) : Product {
        return productRepository.findById(id).get();
    }
    fun save(product: Product):Product {
        val product= productRepository.save(product)
        messageProducerService.sendMessage(product,false,"product.update.updated")
        return product
    }
    fun updateProduct(id: String, product: Product): Product {
        val existingProduct= productRepository.findByIdOrNull(id)
        existingProduct?.category = product.category
        existingProduct?.name =product.name
        return productRepository.save(existingProduct!!)
    }
    fun getProducts(): List<Product> {
        return productRepository.findAll()
    }
    fun getProductsByCategory(categoryId: Long): List<Product> {
        return productRepository.getProductByCategory(categoryId)
    }
    fun delete(id: String) {
        val existingProduct = productRepository.findByIdOrNull(id)
        existingProduct ?: ProductNotFoundException("Product with id $id not found")
        productRepository.deleteById(existingProduct?.id!!)
        messageProducerService.sendMessage(existingProduct!!,true,"product.update.deleted")
    }
}